From 79eacdb7e694c7e6681b68006425df3faca51aec Mon Sep 17 00:00:00 2001
From: Miroslav Lichvar <mlichvar@redhat.com>
Date: Mon, 30 Mar 2015 15:13:27 +0200
Subject: cmdmon: fix initialization of allocated reply slots

When allocating memory to save unacknowledged replies to authenticated
command requests, the last "next" pointer was not initialized to NULL.
When all allocated reply slots were used, the next reply could be
written to an invalid memory instead of allocating a new slot for it.

An attacker that has the command key and is allowed to access cmdmon
(only localhost is allowed by default) could exploit this to crash
chronyd or possibly execute arbitrary code with the privileges of the
chronyd process.
---

diff -urNpa s12/cmdmon.c s13/cmdmon.c
--- s12/cmdmon.c	2015-04-10 21:57:00.000000000 +0200
+++ s13/cmdmon.c	2015-04-10 22:22:11.205094119 +0200
@@ -571,6 +571,7 @@ get_more_replies(void)
     for (i=1; i<REPLY_EXTEND_QUANTUM; i++) {
       new_replies[i-1].next = new_replies + i;
     }
+    new_replies[REPLY_EXTEND_QUANTUM - 1].next = NULL;
     free_replies = new_replies;
   }
 }
